README.md
# CNNImageClassifier

A basic convolutional neural network image classifier using Tensorflow, training against the classic CIFAR-10


ChangeLog.txt
[2/15/17] initial architecture
[2/16/17] Learning Rate Increased 1 -> 1.5, Conv4 Filters Increased 64 -> 96, Max Pool Stride Increased 2 -> 3, Lowered Training Requirements
[2/18/17] Learning Rate Exponential Decay Rate Increased, Max Pool 2 Stride/Size Decreased 3 -> 2, Conv 1 Stride increased 1 -> 2 |Decrease Computaions|
[2/20/17] Convolutional Layers Weight Decay Increased 0 -> 0.004, Added 40% Dropout to Fully Connected Layer (1,2) |Prevent Overfitting|
[2/20/17] Added logic to automatically remove dropout on eval
[2/20/17] Decreased FC1/2 Weight Decay 0.004 -> 0.003, Increased Softmax Weight Decay 0.004 -> 0.006
[2/26/17] Further Decreased Weight Decay 0.003 -> 0.002, Decreased Softmax Weight Decay 0.006 -> 0.004

CNN_test.py
import tensorflow as tf
import numpy as np
import math
import time

import CNN_input
import CNN_model

FLAGS = CNN_input.FLAGS



def image_top_k(image, k):
	return



def main(argv=None):
	return


if __name__ == '__main__':
	tf.app.run(main)

CNN_eval.py
import tensorflow as tf
import numpy as np
import math
import time
from datetime import datetime

import CNN_input
import CNN_model

FLAGS = CNN_input.FLAGS

def eval_once(saver, summary_writer, top_k_op, summary_op):
	with tf.Session() as sess:
		check_point = tf.train.get_checkpoint_state(FLAGS.train_dir)
		if check_point and check_point.model_checkpoint_path:
			saver.restore(sess, check_point.model_checkpoint_path)
			global_step = check_point.model_checkpoint_path.split('/')[-1].split('-')[-1]
		else:
			raise ValueError("No checkpoint file found")

		coord = tf.train.Coordinator()
		threads = []
		try:
			for queue_runner in tf.get_collection(tf.GraphKeys.QUEUE_RUNNERS):
				threads.extend(queue_runner.create_threads(sess, coord=coord, daemon=True, start=True))

			num_iter = int(math.ceil(FLAGS.num_examples / FLAGS.batch_size))
			true_count = 0
			total_sample_count = num_iter * FLAGS.batch_size
			step = 0
			while step < num_iter:
				predictions = sess.run([top_k_op])
				true_count += np.sum(predictions)
				step += 1
				if step % 10 == 0:
					print("%.2f%%" % ((step / num_iter) * 100))

			print(str(true_count) + ' ' + str(total_sample_count) + ' ' + str(step))
			precision = true_count / total_sample_count
			print('%s: precision @ 3 = %.3f' % (datetime.now(), precision))

			summary = tf.Summary()
			summary.ParseFromString(sess.run(summary_op))
			summary.value.add(tag='Precision @ 3', simple_value=precision)
			summary_writer.add_summary(summary, global_step)

			coord.request_stop()
			coord.join(threads, stop_grace_period_secs=10)
		except Exception as e:
			print(e)
			coord.request_stop(e)

def evaluate(run_once):
	with tf.Graph().as_default() as graph:
		images, labels = CNN_input.construct_inputs(True, False)

		logits = CNN_model.inference(False, images)

		top_k_op = tf.nn.in_top_k(logits, labels, 3)

		variable_averages = tf.train.ExponentialMovingAverage(CNN_model.MOVING_AVERAGE_DECAY)
		variables_to_restore = variable_averages.variables_to_restore()
		saver = tf.train.Saver(variables_to_restore)

		summary_op = tf.summary.merge_all()
		summary_writer = tf.summary.FileWriter(FLAGS.eval_dir, graph)

		while True:
			eval_once(saver, summary_writer, top_k_op, summary_op)
			if run_once:
				break
			time.sleep(FLAGS.eval_interval_secs)

def main(argv=None):
	CNN_input.extract()
	if (tf.gfile.Exists(FLAGS.eval_dir)):
		tf.gfile.DeleteRecursively(FLAGS.eval_dir)
	tf.gfile.MakeDirs(FLAGS.eval_dir)
	evaluate(False)

if __name__ == '__main__':
	tf.app.run()


CNN_input.py
import os
import tarfile

import tensorflow as tf

DIRECTORY = os.path.dirname(__file__)

FLAGS = tf.app.flags.FLAGS

# Basic model parameters
tf.app.flags.DEFINE_string('data_dir', os.path.join(DIRECTORY, "data", "cifar10_data"),
						   """Path to CIFAR-10 data directory""")

tf.app.flags.DEFINE_string('train_dir', os.path.join(DIRECTORY, "data", "cifar10_train"),
							"""Directory where to write event logs """
							"""and checkpoint.""")

tf.app.flags.DEFINE_string('eval_dir', os.path.join(DIRECTORY, "data", "cifar10_eval"),
						   """Directory where to write event logs.""")

tf.app.flags.DEFINE_integer('batch_size', 128,
							"""Number of images to process in a batch.""")

tf.app.flags.DEFINE_integer('max_steps', 100000,
							"""Number of batches to run.""")

tf.app.flags.DEFINE_boolean('log_device_placement', False,
							"""Whether to log device placement.""")

tf.app.flags.DEFINE_boolean('report_layer_shapes', True,
							"""Whether to print shapes of network layers""")

tf.app.flags.DEFINE_string('eval_data', 'test',
                           """Either 'test' or 'train_eval'.""")

tf.app.flags.DEFINE_integer('eval_interval_secs', 60 * 3,
                            """How often to run the eval.""")

tf.app.flags.DEFINE_integer('num_examples', 10000,
                            """Number of examples to run during evaluation.""")

#Global Constants
LABEL_BYTES = 1
IMAGE_SIZE = 32
IMAGE_CHANNEL_DEPTH = 3
IMAGE_BYTES = IMAGE_SIZE * IMAGE_SIZE * IMAGE_CHANNEL_DEPTH
RECORD_BYTES = LABEL_BYTES + IMAGE_BYTES

NUM_EXAMPLES_PER_EPOCH_FOR_TRAIN = 4000
NUM_EXAMPLES_PER_EPOCH_FOR_EVAL = 300
MIN_FRACTION_EXAMPLES_QUEUED = 0.4

NUM_CLASSES = 10

FILE_NAME = 'cifar-10-binary.tar.gz'
BATCHES_BIN_FOLDER = 'cifar-10-batches-bin'
BATCH_FILE_FORMAT = 'data_batch_%d.bin'


# Extracts CIFAR-10 Image Data
def extract():
	dest_directory = FLAGS.data_dir

	if (os.path.exists(os.path.join(dest_directory, "cifar-10-batches-bin"))):
		return

	if (not os.path.exists(dest_directory)):
		raise ValueError('Data Directory "' + dest_directory + '" doest not exist!')


	filepath = os.path.join(dest_directory, FILE_NAME)

	if (os.path.exists(filepath)):
		tarfile.open(filepath, 'r:gz').extractall(dest_directory)
	else:
		raise ValueError(FILE_NAME + " does not exist!")


# Reads file from CIFAR-10 filename queue
def read_file(filename_queue):
	reader = tf.FixedLengthRecordReader(record_bytes=RECORD_BYTES)
	key, value = reader.read(filename_queue)

	record_bytes = tf.decode_raw(value, tf.uint8)

	label = tf.cast(tf.strided_slice(record_bytes, [0], [LABEL_BYTES], strides=[1]), tf.int32)

	depth_major = tf.reshape(tf.strided_slice(record_bytes, [LABEL_BYTES], [RECORD_BYTES], strides=[1]),
							 [IMAGE_CHANNEL_DEPTH, IMAGE_SIZE, IMAGE_SIZE])

	uint8image = tf.transpose(depth_major, [1, 2, 0])

	return uint8image, label


# Generates Batch
def _generate_image_and_label_batch(image, label, min_queue_examples, batch_size, shuffle):
	num_preprocess_threads = 16
	if (shuffle):
		images, label_batch = tf.train.shuffle_batch(
			[image, label],
			batch_size=batch_size,
			num_threads=num_preprocess_threads,
			capacity=min_queue_examples + 3 * batch_size,
			min_after_dequeue=min_queue_examples)
	else:
		images, label_batch = tf.train.batch(
			[image, label],
			batch_size=batch_size,
			num_threads=num_preprocess_threads,
			capacity=min_queue_examples + 3 * batch_size)

	tf.summary.image('images', images)

	return images, tf.reshape(label_batch, [batch_size])


def construct_inputs(is_evaluation_inputs, shuffle):
	if (is_evaluation_inputs):
		filenames = [os.path.join(FLAGS.data_dir, BATCHES_BIN_FOLDER, 'test_batch.bin')]
		num_examples_per_epoch = NUM_EXAMPLES_PER_EPOCH_FOR_EVAL
	else:
		filenames = [os.path.join(FLAGS.data_dir, BATCHES_BIN_FOLDER, (BATCH_FILE_FORMAT % i)) for i in range(1, 6)]
		num_examples_per_epoch = NUM_EXAMPLES_PER_EPOCH_FOR_TRAIN

	for file in filenames:
		if (not tf.gfile.Exists(file)):
			raise ValueError('Failed to find file: ' + file)

	filename_queue = tf.train.string_input_producer(filenames)

	uint8image, label = read_file(filename_queue)

	float_image = tf.cast(uint8image, tf.float32)

	# Random distortions to reduce training overfitting
	if (not is_evaluation_inputs):
		float_image = tf.image.random_flip_left_right(float_image)
		float_image = tf.image.random_brightness(float_image, max_delta=63)
		float_image = tf.image.random_contrast(float_image, lower=0.2, upper=1.8)

	finalized_image = tf.image.per_image_standardization(float_image)
	finalized_image.set_shape([IMAGE_SIZE, IMAGE_SIZE, IMAGE_CHANNEL_DEPTH])
	label.set_shape([1])

	min_queue_examples = int(num_examples_per_epoch * MIN_FRACTION_EXAMPLES_QUEUED)

	return _generate_image_and_label_batch(finalized_image,
										   label,
										   min_queue_examples,
										   FLAGS.batch_size,
										   shuffle=shuffle)


CNN_train.py
from datetime import datetime
import time

import numpy as np
import tensorflow as tf

import CNN_input
import CNN_model

FLAGS = CNN_input.FLAGS

LOGGING_STEP_INTERVAL = 3
EST_TIME_SMOOTHING = 25

def train():
	with (tf.Graph().as_default()):
		global_step = tf.contrib.framework.get_or_create_global_step()

		images, labels = CNN_input.construct_inputs(False, True)

		logits = CNN_model.inference(True, images)

		loss = CNN_model.loss(logits, labels)

		train_op = CNN_model.train(loss, global_step)

		# Logs loss and runtime
		class _LoggerHook(tf.train.SessionRunHook):

			def begin(self):
				self._step = -1
				self._estimated_time_array = []
				for i in range(EST_TIME_SMOOTHING):
					self._estimated_time_array.append(0)

			def before_run(self, run_context):
				self._step += 1
				self._start_time = time.time()
				return tf.train.SessionRunArgs(loss)

			def get_estimated_completion(self):
				total = 0
				for i in range(EST_TIME_SMOOTHING):
					total += self._estimated_time_array[i]
				return (total / EST_TIME_SMOOTHING)

			def after_run(self, run_context, run_values):
				duration = time.time() - self._start_time
				loss_value = run_values.results

				self._estimated_time_array[self._step % EST_TIME_SMOOTHING] = float(duration * (FLAGS.max_steps - self._step)) / (60.0 * 60.0)

				if (self._step % LOGGING_STEP_INTERVAL == 0):
					log_value = '%s: step %d, loss = %.2f (%.1f examples/sec; %.3f sec/batch; %.2f est. hours to completion)' \
								% (datetime.now(), self._step, loss_value, FLAGS.batch_size / duration, float(duration), self.get_estimated_completion())
					print(log_value)
					return (log_value)

		with tf.train.MonitoredTrainingSession(
			checkpoint_dir=FLAGS.train_dir,
			hooks=[tf.train.StopAtStepHook(last_step=FLAGS.max_steps),
				   tf.train.NanTensorHook(loss),
				   _LoggerHook()],
			config=tf.ConfigProto(
				log_device_placement=FLAGS.log_device_placement)) as mon_sess:
			while (not mon_sess.should_stop()):
				mon_sess.run(train_op)


def main(argv=None):
	CNN_input.extract()

	if (tf.gfile.Exists(FLAGS.train_dir)):
		tf.gfile.DeleteRecursively(FLAGS.train_dir)
	tf.gfile.MakeDirs(FLAGS.train_dir)

	train()


if __name__ == "__main__":
	tf.app.run()


CNN_model.py
import tensorflow as tf
import numpy as np

import CNN_input

FLAGS = CNN_input.FLAGS

# Global Constants
IMAGE_SIZE = CNN_input.IMAGE_SIZE
IMAGE_CHANNEL_DEPTH = CNN_input.IMAGE_CHANNEL_DEPTH
NUM_EXAMPLES_PER_EMPOCH_FOR_TRAIN = CNN_input.NUM_EXAMPLES_PER_EPOCH_FOR_TRAIN
NUM_EXAMPLES_PER_EMPOCH_FOR_EVAL = CNN_input.NUM_EXAMPLES_PER_EPOCH_FOR_EVAL
NUM_CLASSES = CNN_input.NUM_CLASSES

# Model Constants

# Convolution Layer 1
CONV_1_FILTERS = 64
CONV_1_FILTER_SZ = 5
CONV_1_STRIDE = 2

# Convolution Layer 2
CONV_2_FILTERS = 64
CONV_2_FILTER_SZ = 5
CONV_2_STRIDE = 1

# Max Pool 1
MP1_KSIZE = 2
MP1_STRIDE = 2

# Convolution Layer 3
CONV_3_FILTERS = 96
CONV_3_FILTER_SZ = 3
CONV_3_STRIDE = 1

# Convolution Layer 4
CONV_4_FILTERS = 128
CONV_4_FILTER_SZ = 3
CONV_4_STRIDE = 1

# Max Pool 2
MP2_KSIZE = 2
MP2_STRIDE = 2

# Fully Connected Layer 1
FC1_W_STDDEV = 0.04
FC1_W_DECAY = 0.002
FC1_BIAS_INIT = 0.1
FC1_DROPOUT = 0.4

# Fully Connected Layer 2
FC2_SIZE = 384
FC2_W_STDDEV = 0.04
FC2_W_DECAY = 0.002
FC2_BIAS_INIT = 0.1
FC2_DROPOUT = 0.4

# Fully Connected Softmax Layer 3
FC3_SIZE = 192
FC3_W_STDDEV = 1/FC3_SIZE
FC3_W_DECAY = 0.004
FC3_BIAS_INIT = 0.0

# Training Constants
MOVING_AVERAGE_DECAY = 0.9999
NUM_EPOCHS_PER_DECAY = 25.0
LEARNING_RATE_DECAY_FACTOR = 0.996
INITIAL_LEARNING_RATE = 0.15
NUM_BATCHES_PER_EPOCH = NUM_EXAMPLES_PER_EMPOCH_FOR_TRAIN / FLAGS.batch_size
DECAY_STEPS = int(NUM_BATCHES_PER_EPOCH * NUM_EPOCHS_PER_DECAY)


def _weight_variable_with_decay(name, shape, stddev, weight_decay=None):
	initial = tf.truncated_normal(name=name, shape=shape, stddev=stddev)

	tf_var = tf.Variable(initial)

	if (weight_decay is not None):
		wd = tf.multiply(tf.nn.l2_loss(tf_var), weight_decay, name='weight_loss')
		tf.add_to_collection('losses', wd)

	return tf_var


def _bias_variable(init, shape):
	initial = tf.constant(init, shape=shape, name='biases')
	return tf.Variable(initial)


def _activation_summary(x):
	tensor_name = x.op.name
	tf.summary.histogram(tensor_name + '/activations', x)
	tf.summary.scalar(tensor_name + '/sparsity', tf.nn.zero_fraction(x))


def _build_convolution_layer(name, source_layer, filter_size, filter_depth, num_filters, stride, padding):
	with tf.variable_scope(name):
		kernel = _weight_variable_with_decay('weights',
											 shape=[filter_size, filter_size, filter_depth, num_filters],
											 stddev=5e-2,
											 weight_decay=0.003)
		conv = tf.nn.conv2d(source_layer, kernel, stride, padding=padding)
		biases = _bias_variable(0.1, [num_filters])
		pre_activation = tf.nn.bias_add(conv, biases)
		layer = tf.nn.elu(pre_activation, name=name)
		_activation_summary(layer)

	return layer


def _build_fully_connected_layer(name, source_layer, input_size, output_size, weight_stddev, weight_decay, bias_init, dropout):
	with tf.variable_scope(name) as scope:
		weights = _weight_variable_with_decay('weights',
											  shape=[input_size, output_size],
											  stddev=weight_stddev,
											  weight_decay=weight_decay)
		biases = _bias_variable(bias_init, [output_size])
		fully_con = tf.nn.relu(tf.matmul(source_layer, weights) + biases, name=name)

		if dropout is not None:
			fully_con = tf.nn.dropout(fully_con, 1 - dropout, name=name)

		_activation_summary(fully_con)

		return fully_con

def _print_tensor_shapes(tensor_list):
	if (FLAGS.report_layer_shapes):
		max_len = 0
		for tensor in tensor_list:
			name_len = len(tensor.name.split(':')[0].split('/')[0])
			if (name_len > max_len):
				max_len = name_len

		for tensor in tensor_list:
			name = tensor.name.split(':')[0].split('/')[0]
			print(name + " shape: " + (" "*(max_len - len(name))) + str(tensor.get_shape()))
	return

def inference(is_training, images):
	# Convolution Layer 1
	conv1 = _build_convolution_layer('conv1', images, CONV_1_FILTER_SZ, IMAGE_CHANNEL_DEPTH, CONV_1_FILTERS, [1, CONV_1_STRIDE, CONV_1_STRIDE, 1], 'SAME')

	# Convolution Layer 2
	conv2 = _build_convolution_layer('conv2', conv1, CONV_2_FILTER_SZ, CONV_1_FILTERS, CONV_2_FILTERS, [1, CONV_2_STRIDE, CONV_2_STRIDE, 1], 'SAME')

	# Max Pool 1
	max_pool1 = tf.nn.max_pool(name='pool1', value=conv2, ksize=[1, MP1_KSIZE, MP1_KSIZE, 1], strides=[1, MP1_STRIDE, MP1_STRIDE, 1], padding='SAME')

	# Convolution Layer 3
	conv3 = _build_convolution_layer('conv3', max_pool1, CONV_3_FILTER_SZ, CONV_2_FILTERS, CONV_3_FILTERS, [1, CONV_3_STRIDE, CONV_3_STRIDE, 1], 'SAME')

	# Convolution Layer 4
	conv4 = _build_convolution_layer('conv4', conv3, CONV_4_FILTER_SZ, CONV_3_FILTERS, CONV_4_FILTERS, [1, CONV_4_STRIDE, CONV_4_STRIDE, 1], 'SAME')

	# Max Pool 2
	max_pool2 = tf.nn.max_pool(name='pool2', value=conv4, ksize=[1, MP2_KSIZE, MP2_KSIZE, 1], strides=[1, MP2_STRIDE, MP2_STRIDE, 1], padding='SAME')
	max_pool2_shape = max_pool2.get_shape()

	# Average Pool
	avg_pool = tf.nn.avg_pool(name='avg_pool', value=max_pool2, ksize=[1, max_pool2_shape[1], max_pool2_shape[2], 1], strides=[1, 1, 1, 1], padding='VALID')

	# Reshape Into Depth
	reshape = tf.reshape(avg_pool, [FLAGS.batch_size, -1])

	#Dropouts
	if (is_training):
		fc1_dropout = FC1_DROPOUT
		fc2_dropout = FC2_DROPOUT
	else:
		fc1_dropout = None
		fc2_dropout = None

	# Fully Connected
	fc1 = _build_fully_connected_layer('fc1', reshape, CONV_4_FILTERS, FC2_SIZE, weight_stddev=FC1_W_STDDEV, weight_decay=FC1_W_DECAY, bias_init=FC1_BIAS_INIT, dropout=fc1_dropout)

	# Fully Connected
	fc2 = _build_fully_connected_layer('fc2', fc1, FC2_SIZE, FC3_SIZE, weight_stddev=FC2_W_STDDEV, weight_decay=FC2_W_DECAY, bias_init=FC2_BIAS_INIT, dropout=fc2_dropout)

	# Softmax
	with tf.variable_scope('softmax_linear') as scope:
		weights = _weight_variable_with_decay('weights',
											  [FC3_SIZE, NUM_CLASSES],
											  stddev=FC3_W_STDDEV,
											  weight_decay=FC3_W_DECAY)
		biases = _bias_variable(FC3_BIAS_INIT, [NUM_CLASSES])
		softmax_linear = tf.add(tf.matmul(fc2, weights), biases, name=scope.name)
		_activation_summary(softmax_linear)

	_print_tensor_shapes([images, conv1, conv2, max_pool1, conv3, conv4, max_pool2, avg_pool, reshape, fc1, fc2, softmax_linear])

	return softmax_linear


def loss(logits, labels):
	labels = tf.cast(labels, tf.int64)
	cross_entropy = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits, name='cross_entropy_per_example')
	cross_entropy_mean = tf.reduce_mean(cross_entropy, name='cross_entropy')
	tf.add_to_collection('losses', cross_entropy_mean)

	correct_prediction = tf.equal(labels, tf.argmax(logits, 1))
	accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
	tf.summary.scalar('accuracy', accuracy)

	return tf.add_n(tf.get_collection('losses'), name='total_loss')


def _add_loss_summeries(total_loss):
	loss_averages = tf.train.ExponentialMovingAverage(0.9, name='avg')
	losses = tf.get_collection('losses')
	loss_averages_op = loss_averages.apply(losses + [total_loss])

	for loss in losses + [total_loss]:
		tf.summary.scalar(loss.op.name + ' (raw)', loss)
		tf.summary.scalar(loss.op.name, loss_averages.average(loss))

	return loss_averages_op


def train(total_loss, global_step):
	learning_rate = tf.train.exponential_decay(INITIAL_LEARNING_RATE,
											   global_step,
											   DECAY_STEPS,
											   LEARNING_RATE_DECAY_FACTOR,
											   staircase=True)
	tf.summary.scalar('learning_rate', learning_rate)

	loss_averages_op = _add_loss_summeries(total_loss)

	with tf.control_dependencies([loss_averages_op]):
		optimizer = tf.train.GradientDescentOptimizer(learning_rate)
		gradients = optimizer.compute_gradients(total_loss)

	apply_gradient_op = optimizer.apply_gradients(gradients, global_step=global_step)

	for var in tf.trainable_variables():
		tf.summary.histogram(var.op.name, var)

	for grad, var in gradients:
		if (grad is not None):
			tf.summary.histogram(var.op.name + '/gradients', grad)

	variables_averages = tf.train.ExponentialMovingAverage(MOVING_AVERAGE_DECAY, global_step)
	variables_averages_op = variables_averages.apply(tf.trainable_variables())

	with tf.control_dependencies([apply_gradient_op, variables_averages_op]):
		train_op = tf.no_op(name='train')

	return train_op
